### DISCUSSION TITLE
Reading files in API routes (Next.js)

### DISCUSSION THREAD

**Discussion by sebpowell** [vercel/vercel#6100] [Help]
I have a Next.js project where I'm trying to load .mjml files in the /pages/api routes.

This is my code:

```
export default function handler(req, res) {
  if (req.method === "POST") {
    const { email } = JSON.parse(req.body);
    
    const mjMail = fs.readFileSync(join(__dirname, "public/_files", `${email}.mjml`), "utf8");
    
    const { html, errors } = mjml2html(mjMail, {
      filePath: "public/_files",
    });

    res.status(200).json({ html });
  } else {
  }
}
```

The directory structure is as follows:

```
pages/
-- api/compile.js
public/
-- _files/
```

It works locally, but when I deploy to Vercel, the files aren't properly located by the function and I get an error – e.g:

```
2021-04-12T09:30:27.603Z	61fb97f1-3b8a-4b4b-b554-a2aa154bc84d	ERROR	Error: ENOENT: no such file or directory, open '/_files/verification_email.mjml'
    at Object.openSync (fs.js:476:3)
    at Object.readFileSync (fs.js:377:35)
    at handler (/var/task/.next/server/pages/api/compile.js:127:23)
    at apiResolver (/var/task/node_modules/next/dist/next-server/server/api-utils.js:8:7)
    at processTicksAndRejections (internal/process/task_queues.js:93:5)
    at async Server.handleApiRequest (/var/task/node_modules/next/dist/next-server/server/next-server.js:67:462)
    at async Object.fn (/var/task/node_modules/next/dist/next-server/server/next-server.js:59:492)
    at async Router.execute (/var/task/node_modules/next/dist/next-server/server/router.js:25:67)
    at async Server.run (/var/task/node_modules/next/dist/next-server/server/next-server.js:69:1042)
    at async Server.handleRequest (/var/task/node_modules/next/dist/next-server/server/next-server.js:34:504) {
  errno: -2,
  syscall: 'open',
  code: 'ENOENT',
  path: '/_files/verification_email.mjml'
}
```

I followed the instructions here: https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions and also included the vercel.json file, as follows:

```
{
  "functions": {
    "pages/api/compile.js": {
      "includeFiles": "public/_files/**"
    }
  }
}
```

Any pointers on where I'm going wrong would be greatly appreciated.

**Comment by wiiiimm** [comment]
@sebpowell The example you are referred to is for [Node.js runtime](https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions#node.js). 

As per https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions#next.js, 

> At present, it is not possible to import files in a Next.js SSR page or function given [this issue](https://github.com/vercel/next.js/issues/8251).

**Reply by sebpowell** [reply]
@williamli – thanks for getting back to me!

So just to clarify, I would need to switch over to a Node.js runtime in order for this to work? 

If so, I'm not 100% clear on what I need to do – the only resource I can find relates to creating a custom server (https://nextjs.org/docs/advanced-features/custom-server) – is that what I need to do?

**Comment by wiiiimm** [comment]
Custom server cannot be deployed on Vercel.
To use the Node.js runtime, you can follow the following guides:
https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-version <https://vercel.com/docs/runtimes#official-runtimes/node-js/node-js-version>
https://vercel.com/docs/runtimes#advanced-usage/technical-details/including-additional-files <https://vercel.com/docs/runtimes#advanced-usage/technical-details/including-additional-files>
https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions#next.js <https://vercel.com/support/articles/how-can-i-use-files-in-serverless-functions#next.js>

**Reply by sebpowell** [reply]
@williamli – thank you – as far as I can tell from the documentation, Node.js is the default runtime?

In the 'Advanced' section, it says:

"In order to use this Runtime, no configuration is needed. You only need to create a file inside the api directory."

So if I'm understanding this correctly, the only steps I need to take are:

1.) Create a custom server (as described here: https://nextjs.org/docs/advanced-features/custom-server)

2.) Move the app to a different provider (e.g. Heroku, as Vercel doesn't support custom servers)

Is that right?

**Comment by wiiiimm** [comment]
No, you just have to move your serverless api out from /pages/api folder into /api folder. 

To test locally, instead of using `next dev` you will need to use `vercel dev`.

**Reply by wiiiimm** [reply]
@sebpowell here is a [PR](https://github.com/sebpowell/lead-pro-prototype/pull/1) on changes you need to make.

https://lead-pro-prototype-apjrz4z4r.vercel-support.app/api/compile

**Reply by sebpowell** [reply]
@williamli – amazing, thank you for your help! That's solved it! :D 

Appreciate you making that PR as well, thank you.

Just one last question (out of interest) – I haven't ever seen any examples / documentation suggesting it was possible to move the /api folder outside of pages, this is the first time I've seen it done. 

If I understand everything correctly, what we've done when making this change is switched the project from a Next.js runtime to a Node.js runtime.

Other than making our files readable, it doesn't seem like there are any differences between the Next.js and the Node.js runtimes. But I'm wondering whether you lose out on things like optimisation?

Essentially my question is; is there any reason why you wouldn't want to use the Node.js runtime?

I hope my question makes sense – thanks again for the stellar support!

**Reply by wiiiimm** [reply]
@sebpowell this usage is documented in the links I shared yesterday: https://vercel.com/docs/runtimes#official-runtimes/node-js.

You are switching the API portion of your project from Next.js's API runtime (uses Node in the background) to using Node.js runtime. Since you are using Node.js directly here, you might find some of Next.js's functions unavailable inside your API functions.

**Reply by sebpowell** [reply]
@williamli – all makes sense, thanks for all the info and for your help!
